### PR TITLE
Prevent login DDoS by enforcing max number of sessions per IP per user

### PR DESCRIPTION
**Pull request by burgerkingeater**
This PR adds a configurable upper limit on the number of sessions per IP per User. This is useful when one misbehaving user is creating too many sessions on the same host, kicking out other users' sessions.

### CODE REVIEW

File: azkaban-common/src/main/java/azkaban/server/session/SessionCache.java [hunk 1]
```diff
@@ -53,12 +56,26 @@
   public SessionCache(final Props props) {
     this.effectiveSessionTimeToLive = props.getLong(ConfigurationKeys.SESSION_TIME_TO_LIVE,
         DEFAULT_SESSION_TIME_TO_LIVE);
+
+    Long maxNumberOfSessions = null;
+    try {
+      maxNumberOfSessions = props.getLong(ConfigurationKeys.MAX_SESSION_NUMBER_PER_IP_PER_USER);
```
Comment (ypadron-in) [inline azkaban-common/src/main/java/azkaban/server/session/SessionCache.java]: Should we modify the getLong() method to return an Optional<Long>? It seems this is the only place this method is used so the refactor should be easy?

File: azkaban-common/src/main/java/azkaban/server/session/SessionCache.java [hunk 2]
```diff
@@ -75,11 +92,42 @@ public long getSessionCount() {
...
+   * @return <code>true</code> The number of cached session sharing the same IP and user equals
+   * or greater than the allowed limit if defined;
+   * <code>false</code> otherwise.
+   */
+  private boolean isViolatingMaxNumberOfSessionPerIpPerUser(final Session session) {
+    if (this.maxNumberOfSessionsPerIpPerUser.isPresent()) {
+      // get sessions sharing the same IP and user
+      final Set<Session> sessionsWithSameIP = this.findSessionsByIP(session.getIp());
+      int duplicateSessionCount = 0;
+      for (final Session sessionByIP : sessionsWithSameIP) {
```
Comment (ypadron-in) [inline azkaban-common/src/main/java/azkaban/server/session/SessionCache.java]: Talked offline about encapsulating this logic in a new method findSessionsByIPByUser or similar name.

File: azkaban-common/src/main/java/azkaban/server/session/SessionCache.java (hunk 1 above)
Comment (djaiswal83) [reply azkaban-common/src/main/java/azkaban/server/session/SessionCache.java]: getLong is too generic as the name suggests. Modifying it to return Optional may not be ideal. A wrapper on top of it is possible.

File: azkaban-common/src/main/java/azkaban/server/session/SessionCache.java (hunk 2 above)
Comment (burgerkingeater) [reply azkaban-common/src/main/java/azkaban/server/session/SessionCache.java]: added more comments to make this piece more understandable.

File: azkaban-common/src/main/java/azkaban/server/session/SessionCache.java (hunk 1 above)
Comment (burgerkingeater) [reply azkaban-common/src/main/java/azkaban/server/session/SessionCache.java]: I agree it's better for getLong to return null instead throwing exception when key is not found, but involves a bigger change, will make it a todo.

File: azkaban-common/src/main/java/azkaban/server/session/SessionCache.java (hunk 2 above)
Comment (burgerkingeater) [reply azkaban-common/src/main/java/azkaban/server/session/SessionCache.java]: updated: extract this piece to a new function.

File: azkaban-common/src/main/java/azkaban/server/session/SessionCache.java [hunk 3]
```diff
@@ -53,12 +57,26 @@
   public SessionCache(final Props props) {
     this.effectiveSessionTimeToLive = props.getLong(ConfigurationKeys.SESSION_TIME_TO_LIVE,
         DEFAULT_SESSION_TIME_TO_LIVE);
+
+    Long maxNumberOfSessions = null;
+    try {
+      maxNumberOfSessions = props.getLong(ConfigurationKeys.MAX_SESSION_NUMBER_PER_IP_PER_USER);
+    } catch (final UndefinedPropertyException exception) {
+      maxNumberOfSessions = null;
```
Comment (ypadron-in) [inline azkaban-common/src/main/java/azkaban/server/session/SessionCache.java]: This variable is null already. Should we replace this line with a log message?

Comment (burgerkingeater) [reply azkaban-common/src/main/java/azkaban/server/session/SessionCache.java]: thanks, removed the initializer.

**Pull request merged**